### PR TITLE
Ping collaborators on Slack on new bounties awaiting approval

### LINKED ISSUE #5: Have bot ping relevant developers regarding issues pending bounty approval
### User Story

As a developer, I want to know when a new issue from my team is pending bounty approval, so that I can react more quickly without having to search for information.

### Description

A better way to address this issue would be to leverage the new GitHub bot (written on the [Probot framework](https://probot.github.io/)) to [listen to issue label changes](https://developer.github.com/v3/issues/events/), specifically when an issue is labeled with `bounty-awaiting-approval`, and then act on it by notifying the [team members from the respective repository](https://developer.github.com/v3/repos/collaborators/#list-collaborators) on [Slack](https://github.com/status-im/status-github-bot/blob/master/scripts/greet-new-contributor.js#L73) (either through a DM or a mention on a given room).

Since GitHub user names don't necessarily match Slack user names, the bot would need to read the mapping from a file (JSON or YAML). The same list from #4 should be reused. Only the implementation is required, the file itself will be filled later on by the team.

Documentation for the bot is available in the [wiki](https://wiki.status.im/GitHub_bot).

### PR DESCRIPTION
**Pull request by ImFeelingDucky**
Fixes #5 

Summary: This adds a new Probot script to listen for new issues labeled `bounty-awaiting-approval`, get a list of collaborators, and send a message on Slack mentioning them and linking to the issue.

### CODE REVIEW

File: scripts/assign-to-bounty-awaiting-for-approval.js [hunk 1]
```diff
@@ -131,20 +131,24 @@ async function assignIssueToBountyAwaitingForApproval (context, robot, assign) {
...
 
-  // Send message to Slack
-  const slackHelper = require('../lib/slack')
-  if (assign) {
-    slackHelper.sendMessage(robot, slackClient, config.slack.notification.room, `Assigned issue to ${approvalColumnName} in ${projectBoardName} project\n${payload.issue.html_url}`)
-  } else {
-    slackHelper.sendMessage(robot, slackClient, config.slack.notification.room, `Unassigned issue from ${approvalColumnName} in ${projectBoardName} project\n${payload.issue.html_url}`)
+  if (!process.env.DRY_RUN_BOUNTY_APPROVAL) {
+    // Send message to Slack
+    const slackHelper = require('../lib/slack')
```
Comment (adambabik) [inline scripts/assign-to-bounty-awaiting-for-approval.js]: Can we move it to the top of the file?

File: scripts/bounty-awaiting-approval-slack-ping.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  robot.log('Connected to bounty-awaiting-approval-slack-ping')
+
+  Slack(robot, (slack) => {
+    robot.log.trace('Connected to Slack')
+
+    checkForNewBounties(robot, slack)
+  })
+}
+
+function checkForNewBounties (robot, slackClient) {
```
Comment (adambabik) [inline scripts/bounty-awaiting-approval-slack-ping.js]: As this function registers on `'issues.labeled'` event, maybe we can rename it to `registerForNewBounties`? Should we unregister from `'issues.labeled'` at some point?

File: scripts/bounty-awaiting-approval-slack-ping.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    robot.log.debug(`bountyAwaitingApprovalSlackPing - ${payload.label.name} doesn't match watched ${watchedLabelName} label. Ignoring`)
+    return null
+  }
+
+  robot.log(`bountyAwaitingApprovalSlackPing - issue #${payload.issue.number} on ${ownerName}/${repoName} was labeled as a bounty awaiting approval. Pinging slack...`)
+
+  const slackCollaborators = getSlackCollaborators(ownerName, repoName, github, robot)
+
+  // Don't actually send Slack messages if we're just doing a dry-run
+  if (process.env.DRY_RUN) {
```
Comment (adambabik) [inline scripts/bounty-awaiting-approval-slack-ping.js]: We also use `DRY_RUN_BOUNTY_APPROVAL`. Is it ok or it should be only `DRY_RUN`?

File: scripts/bounty-awaiting-approval-slack-ping.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    `Would have sent a message on Slack to ${config.slack.notification.room} saying: \n` +
+    `New bounty awaiting approval: [#${payload.issue.number} - ${payload.issue.title}](${payload.issue.html_url})
+@${slackCollaborators.join(', @')}`
+    )
+
+    return null
+  }
+
+  // Send message to Slack
+  const slackHelper = require('../lib/slack')
```
Comment (adambabik) [inline scripts/bounty-awaiting-approval-slack-ping.js]: Should be moved at the top of the file?

File: scripts/bounty-awaiting-approval-slack-ping.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  slackHelper.sendMessage(
+    robot,
+    slackClient,
+    config.slack.notification.room,
+    `New bounty awaiting approval: [#${payload.issue.number} - ${payload.issue.title}](${payload.issue.html_url})
+@${slackCollaborators.join(', @')}`
```
Comment (adambabik) [inline scripts/bounty-awaiting-approval-slack-ping.js]: This is pretty pretty hacky :D Maybe 
```js
${slackCollaborators.map(name => `@${name}`).join(', ')}
```
? A little bit more verbose but easier to move around and reason about.

File: scripts/bounty-awaiting-approval-slack-ping.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+  const watchedLabelName = config['bounty-awaiting-approval-slack-ping']['label-name']
+  if (payload.label.name !== watchedLabelName) {
+    robot.log.debug(`bountyAwaitingApprovalSlackPing - ${payload.label.name} doesn't match watched ${watchedLabelName} label. Ignoring`)
+    return null
+  }
+
+  robot.log(`bountyAwaitingApprovalSlackPing - issue #${payload.issue.number} on ${ownerName}/${repoName} was labeled as a bounty awaiting approval. Pinging slack...`)
+
+  const slackCollaborators = getSlackCollaborators(ownerName, repoName, github, robot)
```
Comment (adambabik) [inline scripts/bounty-awaiting-approval-slack-ping.js]: Shouldn't be called with `await getSlackCollaborators(...)`?

File: .github/github-bot.yml [hunk 7]
```diff
@@ -8,6 +8,9 @@ bounty-project-board:
   label-name: 'bounty-awaiting-approval'
   awaiting-approval-column-name: 'bounty-awaiting-approval'
 
+bounty-awaiting-approval-slack-ping:
+  label-name: 'bounty-awaiting-approval'
```
Comment (PombeirP) [inline .github/github-bot.yml]: I think we're safe reusing `bounty-project-board/label-name` here.

File: scripts/bounty-awaiting-approval-slack-ping.js (hunk 3 above)
Comment (PombeirP) [inline scripts/bounty-awaiting-approval-slack-ping.js]: `slackHelper` already takes care of this logic, so you can get rid of this block

File: scripts/bounty-awaiting-approval-slack-ping.js [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    config.slack.notification.room,
+    `New bounty awaiting approval: [#${payload.issue.number} - ${payload.issue.title}](${payload.issue.html_url})
+@${slackCollaborators.join(', @')}`
+  )
+}
+
+// Get the Slack usernames of the collaborators of this repo.
+// Collaborators should add a mapping of their GitHub to Slack usernames in .github/collaborators.yml
+async function getSlackCollaborators (owner, repo, github, robot) {
+  const gitHubToSlackUsernames = defaultConfig(robot, '.github/collaborators.yml').slack
```
Comment (PombeirP) [inline scripts/bounty-awaiting-approval-slack-ping.js]: I think we're limited by GitHub API to access to one file per repo, so we'll unfortunately need to add the contributors map to the `github-bot.yml` file.

File: scripts/bounty-awaiting-approval-slack-ping.js (hunk 2 above)
Comment (PombeirP) [reply scripts/bounty-awaiting-approval-slack-ping.js]: Unregistering is not required as per the Probot samples. Agree with the naming.

File: scripts/bounty-awaiting-approval-slack-ping.js (hunk 3 above)
Comment (PombeirP) [reply scripts/bounty-awaiting-approval-slack-ping.js]: This code needs to go away, since it is repeating logic that is already in `slack.js`.

**Comment by PombeirP** [comment]
@adambabik regarding your questions above:
1. When the bot is installed on an organization, you can select which repos it has access to (and therefore which events it will receive). So the bot code doesn't need to care about it.
2. Maybe we could use the collaborators list to only list those who have mismatched GH/Slack usernames? So for those collaborators who share the same name, we'd just default to the GH username and ping them on Slack using that.

File: scripts/bounty-awaiting-approval-slack-ping.js (hunk 5 above)
Comment (ImFeelingDucky) [reply scripts/bounty-awaiting-approval-slack-ping.js]: Thank you for calling me out on that :laughing: that's much better.

**Comment by ImFeelingDucky** [comment]
Awesome, thanks for the feedback. I have made requested changes.

Also, regarding 2., I thought that was a great idea, so slightly refactored so that GitHub usernames are assumed to also be Slack usernames unless there is a specific entry for that GitHub username in `collaborators.slack` in `github-bot.yml`.

**Comment by adambabik** [comment]
@PombeirP Regarding (2), we can also introduce a special command to identify myself in Probot and provide my Github username. It can keep the mapping persistent. Defaulting to Slack username being also Github username may be dangerous as it can ping someone not within the organization unless this condition will also be checked.

**Comment by PombeirP** [comment]
@adambabik In the end, the most maintenance-free method might be to just have the bot load up the list of active Slack members, lookup the GitHub username field and build the map in memory. That way we're guaranteed that it will always be up-to-date. The problem with using the config file for that is that it is read-only for the bot.

File: scripts/bounty-awaiting-approval-slack-ping.js [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+const Slack = require('probot-slack-status')
+const defaultConfig = require('../lib/config')
+const slackHelper = require('../lib/slack')
+
+module.exports = (robot) => {
+  robot.log('Connected to bounty-awaiting-approval-slack-ping')
+
+    registerForNewBounties(robot, undefined)
+  Slack(robot, (slack) => {
```
Comment (adambabik) [inline scripts/bounty-awaiting-approval-slack-ping.js]: Something wrong with indentation.

**Comment by adambabik** [comment]
@PombeirP by `lookup the GitHub username field` you mean Github username field on Slack profile? I think it's a good idea and maintenance free.

**Comment by PombeirP** [comment]
@adambabik I created a new bounty candidate for that here: https://github.com/status-im/status-github-bot/issues/12. @ImFeelingDucky feel free to tackle it if you're so inclined!

**Comment by PombeirP** [comment]
Hey @ImFeelingDucky, I decided to go ahead and implement a mem-cache of GitHub->Slack IDs that you can leverage in your script. It is in the latest `master`.

**Comment by ImFeelingDucky** [comment]
@PombeirP sweet :+1: I didn't want to take on #12 yet because I already have 2 unclosed bounties, so that's great. I'll update this PR to implement using that in-memory list soon.

**Comment by PombeirP** [comment]
Hey @ImFeelingDucky, you may have noticed that I ended up implementing the last remaining recommendations to this PR, just because we were so close to the finish line, and this functionality is important to us. I'd still like it for you to receive the bounty, which you fully deserve. Just let me know and we can work together to make that happen.

**Pull request closed without merging**